cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/NamespaceSourceControlMetadataRefreshService.java

```diff
@@ -0,0 +1,314 @@
+    this.lastRefreshTime = new AtomicLong(0);
+    this.running = new AtomicBoolean(false);
+    this.batchSize = cConf.getInt(AppFabric.STREAMING_BATCH_SIZE);
+
+    // Adding a jitter (a random value between 0 and 10) to the runInterval because if services for all
+    // the namespaces ran at the same time, it may create availability issues in task workers.
+    long interval =
+        cConf.getLong(SourceControlManagement.METADATA_REFRESH_INTERVAL_SECONDS);
+    if (interval > 0) {
+      Random random = new Random();
+      this.runInterval = interval + random.nextInt(
+          MAX_JITTER_SECONDS + 1);
+    } else {
+      this.runInterval = interval;
+    }
+  }
+
+  private RepositorySourceControlMetadataStore getRepoSourceControlMetadataStore(
+      StructuredTableContext context) {
+    return RepositorySourceControlMetadataStore.create(context);
+  }
+
+  private NamespaceSourceControlMetadataStore getNamespaceSourceControlMetadataStore(
+      StructuredTableContext context) {
+    return NamespaceSourceControlMetadataStore.create(context);
+  }
+
+  @Override
+  protected final ScheduledExecutorService executor() {
+    executor = Executors.newSingleThreadScheduledExecutor(
+        Threads.createDaemonThreadFactory("source-control-metadata-refresh-service-for-namespace-"
+            + namespaceId.getNamespace()));
+    return executor;
+  }
+
+  private RepositoryTable getRepositoryTable(StructuredTableContext context)
+      throws TableNotFoundException {
+    return new RepositoryTable(context);
+  }
+
+
+  // TODO(CDAP-21020): Refactor this function
+  private RepositoryMeta getRepositoryMeta(NamespaceId namespace)
+      throws RepositoryNotFoundException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      RepositoryTable table = getRepositoryTable(context);
+      RepositoryMeta repoMeta = table.get(namespace);
+      if (repoMeta == null) {
+        throw new RepositoryNotFoundException(namespace);
+      }
+
+      return repoMeta;
+    }, RepositoryNotFoundException.class);
+  }
+
+  // TODO(CDAP-21017): Optimize periodic refresh of source control metadata
+  @Override
+  public void runOneIteration() {
+    if (!running.compareAndSet(false, true)) {
+      return;
+    }
+    runRefreshService(false);
+  }
+
+  /**
+   * Triggers a manual refresh of source control metadata for the associated namespace.
+   *
+   * @param forced If true, forces the refresh by skipping the check whether the time elapsed since
+   *               the last refresh is within the buffer time.
+   */
+  public void triggerManualRefresh(boolean forced) {
+    if (!running.compareAndSet(false, true)) {
+      return;
+    }
+    executor().execute(() -> runRefreshService(forced));
+  }
+
+  private void runRefreshService(boolean forced) {
+    running.set(true);
+    Long refreshStartTime = System.currentTimeMillis();
+
+    try {
+
+      if (!namespaceAdmin.exists(namespaceId)) {
```

Comment:
Everything that happens in one run should happens in the same transaction. Using admin will create a different transaction boundary.